src/nvim/insexpand.c

```diff
@@ -2929,10 +2935,19 @@ static void get_complete_info(list_T *what_list, dict_T *retdict)
     }
     if (ret == OK && (what_flag & CI_WHAT_SELECTED)) {
       ret = tv_dict_add_nr(retdict, S_LEN("selected"), selected_idx);
-      win_T *wp = win_float_find_preview();
-      if (wp != NULL) {
-        tv_dict_add_nr(retdict, S_LEN("preview_winid"), wp->handle);
-        tv_dict_add_nr(retdict, S_LEN("preview_bufnr"), wp->w_buffer->handle);
+    }
+  }
+
+  // Support for preview window and preview buffer information
+  if (ret == OK && (what_flag & (CI_WHAT_PREVIEWWIN | CI_WHAT_PREVIEWBUF))) {
+    win_T *wp = win_float_find_preview();
+    if (wp != NULL) {
+      if (what_flag & CI_WHAT_PREVIEWWIN) {
+        ret = tv_dict_add_nr(retdict, S_LEN("preview_winid"), wp->handle);
```

Comment:
What is the actual use-case for this field? The docs are unclear. Is it for showing info about the selected item? Then it should be named something like `info_winid`.

The name `preview_winid` only makes sense if it's coupled to the the `:help preview-window` concept.

Reply:
Explain how it works now When the `info` of `complete-item` is set and `cot+=popup` We use a floating window here `(preview_winid, preview_bufnr)`, these two fields can be obtained in `complete_info()` with `ComplateChanged`, the main purpose is to set the filetype of buffer. lsp usually returns markdown or plaintext. 

The second is that lsp supports [completionitem_resolve request](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem_resolve). Info may be obtained through a second request. Call nvim__complete_set in the asynchronous request callback of lsp to create a floating for the selected item and display info. eg

```lua
vim.api.nvim_create_autocmd('CompleteChanged', {
     callback = function()
                  -- it's better use a debounce timer here
                  local data = vim.fn.complete_info({'selected'})
                  client:requiest('completionitem_resolve', complete_item, function(err, result,...)
                            vim.api.nvim__complete_set(data.selected, {info = result })
                   end, bufnr) 
     end
})
```

Problems noticed. The first selected field of nvim__complete_set_info may be a bit redundant. This should be controlled by the user. Because if C-N/C-P is frequent, a timer is needed here for anti-shake. So the user should know clearly whether the currently selected item and the item before sending the request are the same item when calling nvim_complete_set_info. We do not need this selected parameter for verification.

```lua
                  local data = vim.fn.complete_info({'selected'})
                  client:requiest('completionitem_resolve', complete_item, function(err, result,...)
                           local curdata =  vim.fn.complete_info({'selected'})
                            if curdata.selected ~= data.selected then return end
                           -- a bit redundant ???????????????
                            vim.api.nvim__complete_set(curdata.selected, {info = result })
                   end, bufnr) 
 ````

Reply:
> When the `info` of `complete-item` is set and `cot+=popup` We use a floating window here `(preview_winid, preview_bufnr)`, these two fields can be obtained in `complete_info()`

So is `info_winid` a good name for the field?

Reply:
Yes. This needs to be distinguished from the legacy preview.  just describes the how current works since you missed that PR